### PR TITLE
Futurize - gradients and shell vizualisation

### PR DESCRIPTION
**Pull request by arnaudbore**
clean branch futurize gradients and viz shells see closed PR32

### CODE REVIEW

File: scilpy/samplingscheme/multiple_shell_energy.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+                   verbose=2):
+    """
+    Creates a set of sampling directions on the desired number of shells.
+
+    Parameters
+    ----------
+    nb_shells : the number of shells
+    nb_points_per_shell : list, shape (nb_shells,)
+        A list of integers containing the number of points on each shell.
+    weights : array-like, shep (S, S)
```
Comment (boiteaclou) [inline scilpy/samplingscheme/multiple_shell_energy.py]: typo "shep"

File: scilpy/samplingscheme/multiple_shell_energy.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+            datafile.write("%d\t%f\t%f\t%f\n" %
+                           (s, vects[k, 0], vects[k, 1], vects[k, 2]))
+            k += 1
+    datafile.close()
+
+
+def random_uniform_on_sphere(K):
+    """
+    Creates a set of K pseudo-random unit vectors, following a uniform
+    distribution on the sphere.
```
Comment (boiteaclou) [inline scilpy/samplingscheme/multiple_shell_energy.py]: Missing parameters from docstring

File: scilpy/samplingscheme/multiple_shell_energy.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    vects[:, 1] = np.sin(theta) * np.sin(phi)
+    vects[:, 2] = np.cos(theta)
+
+    return vects
+
+
+def compute_weights(nb_shells, nb_points_per_shell, shell_groups, alphas):
+    """
+    Computes the weights array from a set of shell groups to couple, and
+    coupling weights.
```
Comment (boiteaclou) [inline scilpy/samplingscheme/multiple_shell_energy.py]: Missing parameters from docstring

File: scilpy/samplingscheme/optimize_scheme.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import division
```
Comment (boiteaclou) [inline scilpy/samplingscheme/optimize_scheme.py]: To remove

File: scilpy/samplingscheme/optimize_scheme.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+        Shell index for bvecs in points
+    """
+
+    new_points = points.copy()
+
+    non_b0_pts = points[np.where(shell_idx != -1)]
+
+    it = 0
+    # Assume non-collinearity of non-b0s b-vectors (i.e. Caruyer sampler type)
+    for idx in np.where(shell_idx == -1)[0]:
```
Comment (boiteaclou) [inline scilpy/samplingscheme/optimize_scheme.py]: This can be replaced by numpy indexing new_points[np.where ...] = non_b0_pts

File: scilpy/samplingscheme/optimize_scheme.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+    new_points[non_b0s_mask] = points[non_b0s_mask][ordering_best]
+
+    new_shell_idx = shell_idx.copy()
+    new_shell_idx[non_b0s_mask] = shell_idx[non_b0s_mask][ordering_best]
+
+    return new_points, new_shell_idx
+
+
+def _peak_power(q_scheme, ker_size=10):
+    # Note: np.convolve inverses the filter
```
Comment (boiteaclou) [inline scilpy/samplingscheme/optimize_scheme.py]: Missing docstring

File: scilpy/samplingscheme/optimize_scheme.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+    for idx in np.where(shell_idx == -1)[0]:
+        new_points[idx] = non_b0_pts[it]
+        it += 1
+
+    logging.info('Done adapting b0s for Philips scanner.')
+
+    return new_points, shell_idx
+
+
+def min_duty_cycle_bruteforce(points, shell_idx, bvalues, ker_size=10,
```
Comment (boiteaclou) [inline scilpy/samplingscheme/optimize_scheme.py]: Rename with verb at the beginning (compute_ or other)

File: scilpy/samplingscheme/optimize_scheme.py [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+    new_points = points.copy()
+    new_points[non_b0s_mask] = points[non_b0s_mask][ordering_best]
+
+    new_shell_idx = shell_idx.copy()
+    new_shell_idx[non_b0s_mask] = shell_idx[non_b0s_mask][ordering_best]
+
+    return new_points, new_shell_idx
+
+
+def _peak_power(q_scheme, ker_size=10):
```
Comment (boiteaclou) [inline scilpy/samplingscheme/optimize_scheme.py]: Private name for function ? We should add a verb at the beginning also

File: scilpy/samplingscheme/optimize_scheme.py [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+    pow_z = np.convolve(q_scheme[:, 2], ker, 'full')[:-(ker_size-1)]
+
+    max_pow_x = np.max(pow_x)
+    max_pow_y = np.max(pow_y)
+    max_pow_z = np.max(pow_z)
+
+    return np.max([max_pow_x, max_pow_y, max_pow_z])
+
+
+def bvalue_lin_q(bmin=0.0, bmax=3000.0, nb_of_b_inside=2, exclude_bmin=True,
```
Comment (boiteaclou) [inline scilpy/samplingscheme/optimize_scheme.py]: Add verb

File: scilpy/samplingscheme/optimize_scheme.py [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+                                nb_of_b_inside + 2)**2)
+    if exclude_bmin:
+        b_values = b_values[1:]
+
+    logging.info('b-values linear in q: {}'.format(b_values))
+
+    return b_values
+
+
+def bvalue_lin_b(bmin=0.0, bmax=3000.0, nb_of_b_inside=2, exclude_bmin=True,
```
Comment (boiteaclou) [inline scilpy/samplingscheme/optimize_scheme.py]: Add verb

File: scilpy/samplingscheme/save_scheme.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import division
```
Comment (boiteaclou) [inline scilpy/samplingscheme/save_scheme.py]: To remove

File: scripts/scil_visualize_multi_shell.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    p.add_argument(
+        '--no-sym', action='store_false', dest='sym',
+        help='Disable antipodal symmetry.')
+    p.add_argument(
+        '--no-sphere', action='store_false', dest='sph',
+        help='Disable the rendering of the sphere.')
+    p.add_argument(
+        '--same', action='store_true',
+        help='Use same color for all shell.')
+    p.add_argument(
```
Comment (boiteaclou) [inline scripts/scil_visualize_multi_shell.py]: Same as for sym and sph

File: scripts/scil_visualize_multi_shell.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+
+DESCRIPTION = """
+Vizualisation for sampling schemes.
+Only supports .caru, .txt (Philips), .dir or .dvs (Siemens), .bvecs/.bvals
+and .b (MRtrix).
+"""
+
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(
```
Comment (boiteaclou) [inline scripts/scil_visualize_multi_shell.py]: This arg parser could be made clearer with the help of groups. Also, the ordering of the arguments is inconsistent

File: scripts/scil_visualize_multi_shell.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    proj = args.proj
+    each = args.each
+
+    if not (proj or each):
+        parser.error('Select at least one type of rendering (proj or each).')
+
+    # In no way robust, assume the input is from generate_sampling_scheme.py
+    # For bvec(s)/bval(s)/FSL format, uses bad assumption for Transpose
+
+    # Use
```
Comment (boiteaclou) [inline scripts/scil_visualize_multi_shell.py]: Use ?

File: scripts/scil_generate_sampling_scheme.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+                   action='store_true',
+                   help='Apply duty cycle optimization. '
+                        '\nB-vectors are shuffled to reduce consecutive '
+                        'colinearity in the samples. [%(default)s]')
+    p.add_argument('--b0inter',
+                   dest='b0_every', type=int, default=-1,
+                   help='Interleave a b0 every b0_every. \nNo b0 if 0. '
+                        '\nOnly 1 b0 at beginning if > number of samples '
+                        'or negative. [%(default)s]')
+    p.add_argument('--b0end',
```
[the remark this answers is not in the document]
Comment (arnaudbore) [reply scripts/scil_generate_sampling_scheme.py]: I'll keep that for a future PR.

**Comment by jchoude** [comment]
Were the Caruyer functions merged in Dipy? If so, the ones copied here should be replaced with the Dipy ones.

File: scilpy/viz/sampling_scheme.py [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+        if ofile:
+            window.snapshot(ren, fname=ofile + '_shell_' + str(i) + '.png',
+                            size=ores)
+
+
+def plot_proj_shell(ms, use_sym=True, use_sphere=True, same_color=False,
+                    rad=0.025, opacity=1.0, ofile=None, ores=(300, 300)):
+    """
+    Plot each shell
```
[the remark this answers is not in the document]
Comment (arnaudbore) [reply scilpy/viz/sampling_scheme.py]: You save a file per shell or you save everything together. Maybe it could/should be merged but nothing critical I would say.

**Comment by arnaudbore** [comment]
> Were the Caruyer functions merged in Dipy? If so, the ones copied here should be replaced with the Dipy ones.

I agree but looking at the code in dipy and merging with what we have, does not seem trivial to me. I'll let this for another PR.

File: scilpy/viz/sampling_scheme.py [hunk 17]
```diff
@@ -145,7 +143,7 @@ def plot_proj_shell(ms, use_sym=True, use_sphere=True, same_color=False,
 
 def build_shell_idx_from_bval(bvals, shell_th=50):
     """
-    Plot each shell
+    build_shell_idx_from_bval
```
Comment (boiteaclou) [inline scilpy/viz/sampling_scheme.py]: Bad formatting ? Is it normal ? Same for build_ms_from_shell_idx

File: scilpy/samplingscheme/gen_scheme.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import division
```
Comment (boiteaclou) [inline scilpy/samplingscheme/gen_scheme.py]: Some future import got back on this version, Are they required ?

File: scripts/scil_generate_sampling_scheme.py [hunk 19]
```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+from __future__ import division
```
Comment (boiteaclou) [inline scripts/scil_generate_sampling_scheme.py]: Another future ...

File: scilpy/samplingscheme/multiple_shell_energy.py [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+    S: int
+        Number of shells.
+    Ks: list of ints, len(Ks) = S. Number of points per shell.
+    weights : array-like, shape (S, S)
+        Weighting parameter, control coupling between shells and how this
+        balances.
+
+    Returns
+    -------
+    f: function
```
Comment (boiteaclou) [inline scilpy/samplingscheme/multiple_shell_energy.py]: This returns a float. The function f is evaluated

File: scilpy/samplingscheme/multiple_shell_energy.py [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+        number of shells
+    Ks : list of ints
+        len(Ks) = S. Number of points per shell.
+    weights : array-like, shape (S, S)
+        weighting parameter, control coupling between shells and how this
+        balances.
+
+    Returns
+    -------
+    grad_f: function
```
Comment (boiteaclou) [inline scilpy/samplingscheme/multiple_shell_energy.py]: This also returns a float

File: scilpy/samplingscheme/save_scheme.py [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+        bvecs normalized to 1.
+    shell_idx: numpy.array
+        Shell index for bvecs in points.
+    filename: output file name
+    ------
+    """
+
+    fullfilename, ext = split_name_with_nii(filename)
+    fullfilename = fullfilename + '.caru'
+    f = open(fullfilename, 'w')
```
Comment (boiteaclou) [inline scilpy/samplingscheme/save_scheme.py]: Maybe could be put in another pull request, but, for all file writting operations, we should use the managers for files from python (with open(...) as f)

File: scilpy/samplingscheme/gen_scheme.py (hunk 18 above)
Comment (arnaudbore) [reply scilpy/samplingscheme/gen_scheme.py]: weird I'm fixing this

File: scripts/scil_visualize_multi_shell.py [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+    g1.add_argument(
+        '--dis-proj', action='store_false', dest='enable_proj',
+        help='Disable rendering of the projection supershell.')
+    g1.add_argument(
+        '--plot-shells', action='store_true', dest='plot_shells',
+        help='Enable rendering each shell individually.')
+
+    g2 = p.add_argument_group(title='Rendering options')
+    g2.add_argument(
+        '--same-color', action='store_true', dest='same_color',
```
[the remark this answers is not in the document]
Comment (arnaudbore) [reply scripts/scil_visualize_multi_shell.py]: destination uses underscore.

File: scripts/scil_visualize_multi_shell.py [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+def main():
+    parser = _build_args_parser()
+    args = parser.parse_args()
+    assert_inputs_exist(parser, [args.scheme_file])
+
+    if args.out:
+        possibleOutputPaths = [args.out + '_shell_' + str(i) +
+                               '.png' for i in range(30)]
+        possibleOutputPaths.append(args.out + '.png')
+        assert_outputs_exist(parser, args, possibleOutputPaths)
```
[the remark this answers is not in the document]
Comment (arnaudbore) [reply scripts/scil_visualize_multi_shell.py]: There's not a lot of computation time and it will go quickly to L165. For the output, I split if needed the extension from basename.

File: scripts/scil_visualize_multi_shell.py [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+
+    g1 = p.add_argument_group(title='Enable/Disable renderings')
+    g1.add_argument(
+        '--dis-sphere', action='store_false', dest='enable_sph',
+        help='Disable the rendering of the sphere.')
+    g1.add_argument(
+        '--dis-proj', action='store_false', dest='enable_proj',
+        help='Disable rendering of the projection supershell.')
+    g1.add_argument(
+        '--plot-shells', action='store_true', dest='plot_shells',
```
[the remark this answers is not in the document]
Comment (arnaudbore) [reply scripts/scil_visualize_multi_shell.py]: same -> underscores

File: scilpy/samplingscheme/multiple_shell_energy.py [hunk 26]
```diff
@@ -0,0 +1,305 @@
...
+
+    Parameters
+    ----------
+    nb_shells : the number of shells
+    nb_points_per_shell : list, shape (nb_shells,)
+        A list of integers containing the number of points on each shell.
+    weights : array-like, shape (S, S)
+        weighting parameter, control coupling between shells and how this
+        balances.
+
```
[the remark this answers is not in the document]
Comment (arnaudbore) [reply scilpy/samplingscheme/multiple_shell_energy.py]: Everything is done that way because of https://docs.scipy.org/doc/scipy-0.13.0/reference/generated/scipy.optimize.fmin_slsqp.html

File: scilpy/samplingscheme/save_scheme.py (hunk 22 above)
Comment (GuillaumeTh) [reply scilpy/samplingscheme/save_scheme.py]: It could be nice to have this in this PR.

File: scripts/scil_convert_gradients_mrtrix_to_fsl.py [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter,
+                                description=DESCRIPTION)
+
+    p.add_argument('mrtrix_enc',
+                   help='Gradient directions encoding file. (.b)')
+
+    p.add_argument('fsl_basename',
```
Comment (GuillaumeTh) [inline scripts/scil_convert_gradients_mrtrix_to_fsl.py]: I personally prefer to have bval and bvec argument separated. With this it is mandatory to have .bval and .bvec

File: scripts/scil_generate_sampling_scheme.py [hunk 28]
```diff
@@ -0,0 +1,213 @@
...
+    p.add_argument('--b0inter',
+                   dest='b0_every', type=int, default=-1,
+                   help='Interleave a b0 every b0_every. \nNo b0 if 0. '
+                        '\nOnly 1 b0 at beginning if > number of samples '
+                        'or negative. [%(default)s]')
+    p.add_argument('--b0end',
+                   action='store_true', dest='b0_end',
+                   help='Add a b0 as last sample. [%(default)s]')
+    p.add_argument('--b0value',
+                   dest='b0_value', type=float, default=0.01,
```
Comment (GuillaumeTh) [inline scripts/scil_generate_sampling_scheme.py]: Why 0.01?

File: scripts/scil_generate_sampling_scheme.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (GuillaumeTh) [reply scripts/scil_generate_sampling_scheme.py]: Is it possible to do this now ? It is not a lot of modifications.

File: scripts/scil_generate_sampling_scheme.py [hunk 29]
```diff
@@ -0,0 +1,213 @@
...
+    mrtrix = args.mrtrix
+
+    if not (caru or phil or fsl or siemens or mrtrix):
+        parser.error('Select at least one save format.')
+        return
+
+    if args.verbose:
+        logging.basicConfig(level=logging.DEBUG)
+    else:
+        logging.basicConfig(level=logging.INFO)
```
Comment (GuillaumeTh) [inline scripts/scil_generate_sampling_scheme.py]: I don't like the fact that you always have text displayed even if the verbose version is not activated.

File: scripts/scil_generate_sampling_scheme.py (hunk 28 above)
Comment (arnaudbore) [reply scripts/scil_generate_sampling_scheme.py]: I can put 0.1 doesn't really matter.

File: scripts/scil_generate_sampling_scheme.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (arnaudbore) [reply scripts/scil_generate_sampling_scheme.py]: I don't understand how that's possible.
-1 -> Only 1 b0 at the begining 
0 -> no b0
>0 -> One b0 at the begining and at the end ??
If I don't want one b0 at the end how do I do ?

[the remark this answers is not in the document]
Comment (boiteaclou) [reply scripts/scil_generate_sampling_scheme.py]: I hadn't though of this use case. You're right, only using those ranges of values, it cannot be done. The only way I see it possible is if we can use the negative range of values also, but it's not really optimal nor clear, the other parameter is a better option.

[the remark this answers is not in the document]
Comment (boiteaclou) [reply scripts/scil_generate_sampling_scheme.py]: I hadn't though of this use case. You're right, only using those ranges of values, it cannot be done. The only way I see it possible is if we can use the negative range of values also, but it's not really optimal nor clear, the other parameter is a better option.
